target/linux/lantiq/config-5.15

```diff
@@ -18,6 +18,16 @@ CONFIG_ARCH_USE_QUEUED_RWLOCKS=y
 CONFIG_ARCH_USE_QUEUED_SPINLOCKS=y
 CONFIG_ARCH_WANT_DEFAULT_TOPDOWN_MMAP_LAYOUT=y
 CONFIG_ARCH_WANT_IPC_PARSE_VERSION=y
+CONFIG_BCMA=y
```

Comment:
Only BCMA or SSB should be needed.

I would prefer if you use the OpenWrt kernel package from kmod-bcma or kmod-ssb instead of building this into the kernel. The build scripts would pack the kmod-* packages only into the root fs for this device, if it is build in we ship this on all devices suing this target.

Reply:
Directly including those kmod packages on the makefile will cause the source not to compile:

`mips-openwrt-linux-musl-ld: arch/mips/lantiq/b43-sprom.o: in function `b43_sprom_probe':
arch/mips/lantiq/b43-sprom.c:(.text.unlikely.b43_sprom_probe+0x160): undefined reference to `bcma_arch_register_fallback_sprom'
make[5]: *** [Makefile:1195: vmlinux] Error 1
make[5]: Leaving directory '/home/osboxes/newopenwrt/openwrt-Arcadyan-VRV9510KWAC23/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/linux-5.10.155'
make[4]: *** [Makefile:22: /home/osboxes/newopenwrt/openwrt-Arcadyan-VRV9510KWAC23/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/linux-5.10.155/.modules] Error 2
`

Those variables have dependencies with the added b43_sprom driver

Reply:
As Hauke suggested, the driver for Broadcom's wifi should be built as a package instead of being compiled into the kernel.

Reply:
The `target/linux/lantiq/files/arch/mips/lantiq/b43-sprom.c` will provide a dummy sprom for the cards. I do not like this approach because we have to compile the calibration data into the kernel.

I think the best approach would be to have something like this sprom driver which reads the data from a file instead of the brcm47xx nvram: https://github.com/torvalds/linux/blob/master/drivers/firmware/broadcom/bcm47xx_sprom.c Then you can provide a nvram file in user space and the driver will load and interpret it as a sprom. The bcm47xx_nvram.c code already parses a a nvram. The brcmfmac driver does something similar.
It could parse a file like this: https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/brcm/brcmfmac43340-sdio.meegopad-t08.txt

If you think that is too complicated you could try to move the b43-sprom.c code to brcm or ssb and include it into that module.

Reply:
> The `target/linux/lantiq/files/arch/mips/lantiq/b43-sprom.c` will provide a dummy sprom for the cards. I do not like this approach because we have to compile the calibration data into the kernel.
> 
> I think the best approach would be to have something like this sprom driver which reads the data from a file instead of the brcm47xx nvram: https://github.com/torvalds/linux/blob/master/drivers/firmware/broadcom/bcm47xx_sprom.c Then you can provide a nvram file in user space and the driver will load and interpret it as a sprom. The bcm47xx_nvram.c code already parses a a nvram. The brcmfmac driver does something similar. It could parse a file like this: https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/brcm/brcmfmac43340-sdio.meegopad-t08.txt
> 
> If you think that is too complicated you could try to move the b43-sprom.c code to brcm or ssb and include it into that module.

@hauke Your approach seems interesting, and I will try to get that working. But my first goal is to achieve what you are asking me about not compiling the fallback driver on the kernel. If I move that b43-sprom.c into kmod-brcm or kmod-ssb, won't that affect all devices using that kernel module? I see that even more critical than compiling that in the kernel.

Reply:
@danielhuici see https://github.com/openwrt/openwrt/pull/12427